README.md
# molindo-node-logger

A node.js logger that integrates well with the Molindo infrastructure.

## Features
 - Pretty prints messages in development mode.
 - Prints JSON messages in production mode.
 - Errors are printed to `stderr`, while all other levels are printed to `stdout`.
 - In development only infos, warnings and errors are printed, while in production all levels are printed.
 - If the process encounters an error, it will log it to the console before shutting down. `unhandledRejection` errors thrown from promises are logged as well.
 - Offers an express integration that logs HTTP requests, including the `operationName` in the case of GraphQL requests. Confidential headers like `cookie` or `authorization` are masked.
 - Uses the levels and level values from [logback](https://logback.qos.ch/): `ERROR` (40000), `WARN` (30000), `INFO` (20000), `DEBUG` (10000), `TRACE` (5000).

## Usage

```js
import Logger from 'molindo-node-logger';

const logger = new Logger({service: 'pizza-shop'});
logger.trace('Making a salami pizza …');
logger.debug('Adding salami …');
logger.info('Putting it in the oven …');
logger.warn('Don\'t forget to get it out in time …');
logger.error('Oh no, the pizza is burned!');
```

In production, printed JSON will look like this (except that it's not pretty printed):

```json
{
  "service": "pizza-shop",
  "@timestamp": "2017-10-19T08:26:13.168Z",
  "level": "INFO",
  "level_value": 20000,
  "message": "Pizza is ready!"
}
```

### Express integration

If you're running an express server, you can register the logger middleware to log HTTP requests.

```js
import express from 'express';
import Logger, {createLoggerMiddleware} from 'molindo-node-logger';

const server = express();

const logger = new Logger({service: 'pizza-shop'});
server.use(createLoggerMiddleware({logger}));
```


CHANGELOG.md
# Changelog

## 1.1.1

Fix logging of GraphQL response in case of errors ([@symn](https://github.com/symn) in [#3](https://github.com/molindo/molindo-node-logger/pull/3)).


src/__tests__/index-test.js
import * as index from '../index';

it('exports all necessary modules', () => {
  expect(Object.keys(index)).toEqual(['default', 'createLoggerMiddleware']);
});


src/__tests__/Logger-test.js
import Logger from '../Logger';

describe('Logger', () => {
  const originalStdout = process.stdout.write;
  const originalStderr = process.stderr.write;
  beforeEach(() => {
    process.stdout.write = jest.fn();
    process.stderr.write = jest.fn();
  });

  it('logs plain messages in development', () => {
    const logger = new Logger({
      service: 'pizza-shop',
      colorize: false,
      level: 'TRACE',
      isProduction: false
    });
    logger.trace('making a salami pizza');
    logger.debug('adding salami');
    logger.info('putting it in the oven');
    logger.warn("don't forget to get it out in time");
    logger.error('pizza is burned!');

    expect(process.stdout.write.mock.calls).toEqual([
      ['TRACE: making a salami pizza\n'],
      ['DEBUG: adding salami\n'],
      ['INFO: putting it in the oven\n'],
      ["WARN: don't forget to get it out in time\n"]
    ]);
    expect(process.stderr.write.mock.calls).toEqual([
      ['ERROR: pizza is burned!\n']
    ]);

    logger.destroy();
  });

  it('only logs infos, warnings and errors in development', () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: false});
    logger.trace('making a salami pizza');
    logger.debug('adding salami');
    logger.info('putting it in the oven');
    logger.warn("don't forget to get it out in time");
    logger.error('pizza is burned!');

    expect(process.stdout.write.mock.calls.length).toBe(2);
    expect(process.stderr.write.mock.calls.length).toBe(1);

    logger.destroy();
  });

  it('logs all levels in production', () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: true});
    logger.trace('making a salami pizza');
    logger.debug('adding salami');
    logger.info('putting it in the oven');
    logger.warn("don't forget to get it out in time");
    logger.error('pizza is burned!');

    expect(process.stdout.write.mock.calls.length).toBe(4);
    expect(process.stderr.write.mock.calls.length).toBe(1);

    logger.destroy();
  });

  it('logs json messages in production', () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: true});
    logger.info('pizza is ready!');

    const output = JSON.parse(process.stdout.write.mock.calls[0][0]);
    expect(typeof output['@timestamp']).toBe('string');
    expect(output['@timestamp']).toMatch(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/);
    expect(output.level).toBe('INFO');
    expect(output.level_value).toBe(20000);
    expect(output.message).toBe('pizza is ready!');
    expect(output.service).toBe('pizza-shop');

    logger.destroy();
  });

  it('overloads the log functions to provide a logger name', () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: true});

    logger.trace({
      message: 'Making a salami pizza …',
      name: 'cook'
    });
    logger.info({
      message: 'Pizza can be served now!',
      name: 'cook'
    });

    const stdoutCalls = process.stdout.write.mock.calls.map(call =>
      JSON.parse(call[0])
    );

    expect(stdoutCalls[0]['logger_name']).toBe('cook');
    expect(stdoutCalls[1]['logger_name']).toBe('cook');

    logger.destroy();
  });

  it('throws when no service name is provided', () => {
    expect(() => new Logger()).toThrow(/`service` is mandatory/);
  });

  afterEach(() => {
    process.stdout.write = originalStdout;
    process.stderr.write = originalStderr;
  });
});


src/__tests__/createLoggerMiddleware-test.js
import express from 'express';
import supertest from 'supertest';
import Logger from '../Logger';
import createLoggerMiddleware from '../createLoggerMiddleware';

const createServer = logger => {
  const server = express();

  server.use(createLoggerMiddleware({logger}));
  server.get('/', (req, res) => res.json({success: true}));
  server.get('/500', () => {
    throw new Error('500');
  });
  server.post('/graphql', (req, res) => {
    if (req.body.operationName === 'error') {
      res.status(500);
      res.json({
        errors: [
          {
            message: '401: Unauthorized',
            locations: [{line: 3, column: 3}],
            path: ['pizzas'],
            extensions: {
              statusCode: 401,
              statusText: 'Unauthorized',
              responseText: 'Unauthorized',
              method: 'GET',
              url: 'https://api.example.com/pizzas'
            }
          }
        ],
        data: null
      });
    } else {
      res.json({data: {publicId: '1'}});
    }
  });
  return server;
};

const mockGraphQLPayload = {
  operationName: 'createPizza',
  mutation: `
  mutation createPizza($pizza: PizzaInput!) {
    createPizza(pizza: $pizza) {
      publicId
    }
  }
`,
  variables: {pizza: {toppings: ['salami']}}
};

describe('createLoggerMiddleware', () => {
  const originalStdout = process.stdout.write;
  const originalStderr = process.stderr.write;

  beforeEach(() => {
    process.stdout.write = jest.fn();
    process.stderr.write = jest.fn();
  });

  it('logs plain messages in development', async () => {
    const logger = new Logger({
      service: 'pizza-shop',
      level: 'TRACE',
      isProduction: false,
      colorize: false
    });
    const server = createServer(logger);

    await supertest(server).get('/');
    await supertest(server).get('/404');
    await supertest(server)
      .post('/graphql')
      .send(mockGraphQLPayload);
    await supertest(server)
      .post('/graphql')
      .send({...mockGraphQLPayload, operationName: 'error'});
    await supertest(server).get('/500');

    const stdoutCalls = process.stdout.write.mock.calls.map(call => call[0]);
    expect(stdoutCalls[0]).toMatch(
      /DEBUG: HTTP GET \/ statusCode=200,.*url=\//
    );
    expect(stdoutCalls[1]).toMatch(
      /WARN: HTTP GET \/404 statusCode=404,[\s\S]*url=\/404/
    );
    expect(stdoutCalls[2]).toMatch(
      /DEBUG: HTTP POST \/graphql statusCode=200,.*url=\/graphql/
    );

    const stderrCalls = process.stderr.write.mock.calls.map(call => call[0]);
    expect(stderrCalls[0]).toMatch(
      /ERROR: HTTP POST \/graphql statusCode=500[\s\S]*url=\/graphql/
    );
    expect(stderrCalls[1]).toMatch(
      /ERROR: HTTP GET \/500 statusCode=500[\s\S]*url=\/500/
    );

    logger.destroy();
  });

  it('logs json messages in production', async () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: true});
    const server = createServer(logger);

    await supertest(server).get('/');
    await supertest(server).get('/404');
    await supertest(server)
      .post('/graphql')
      .send(mockGraphQLPayload);
    await supertest(server)
      .post('/graphql')
      .send({...mockGraphQLPayload, operationName: 'error'});
    await supertest(server).get('/500');

    const stdoutCalls = process.stdout.write.mock.calls.map(call =>
      JSON.parse(call[0])
    );

    expect(typeof stdoutCalls[0]['@timestamp']).toBe('string');
    expect(stdoutCalls[0]['@timestamp']).toMatch(
      /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/
    );
    expect(stdoutCalls[0].level).toBe('DEBUG');
    expect(stdoutCalls[0].level_value).toBe(10000);
    expect(stdoutCalls[0].logger_name).toBe('express');
    expect(stdoutCalls[0].message).toBe('HTTP GET /');
    expect(stdoutCalls[0].service).toBe('pizza-shop');

    expect(stdoutCalls[0].meta).toBeTruthy();
    expect(stdoutCalls[0].meta.req).toBeTruthy();
    expect(stdoutCalls[0].meta.req.headers).toBeTruthy();
    expect(stdoutCalls[0].meta.req.method).toBe('GET');
    expect(stdoutCalls[0].meta.req.originalUrl).toBe('/');
    expect(stdoutCalls[0].meta.req.body).toBe(undefined);
    expect(stdoutCalls[0].meta.res).toBeTruthy();
    expect(stdoutCalls[0].meta.res.statusCode).toBe(200);
    expect(typeof stdoutCalls[0].meta.responseTime).toBe('number');

    expect(stdoutCalls[1].level).toBe('WARN');

    expect(stdoutCalls[2].level).toBe('DEBUG');
    expect(stdoutCalls[2].meta.graphql).toEqual({
      operationName: 'createPizza',
      variables: {pizza: {toppings: ['salami']}}
    });

    const stderrCalls = process.stderr.write.mock.calls.map(call =>
      JSON.parse(call[0])
    );
    expect(stderrCalls[0].level).toBe('ERROR');
    expect(stderrCalls[0].meta.res.body).toEqual({
      errors: [
        {
          extensions: {
            method: 'GET',
            responseText: 'Unauthorized',
            statusCode: 401,
            statusText: 'Unauthorized',
            url: 'https://api.example.com/pizzas'
          },
          locations: [{column: 3, line: 3}],
          message: '401: Unauthorized',
          path: ['pizzas']
        }
      ],
      data: null
    });
    expect(stderrCalls[1].level).toBe('ERROR');

    logger.destroy();
  });

  it('masks confidential headers', async () => {
    const logger = new Logger({service: 'pizza-shop', isProduction: true});
    const server = createServer(logger);

    await supertest(server)
      .get('/')
      .set({
        cookie: 'JSESSIONID=1234567890',
        'accept-language': 'en-US,en;q=0.8,de;q=0.6,la;q=0.4',
        authorization: 'Bearer cn389ncoiwuencr',
        accept: 'application/hal+json, application/json',
        'x-requested-with': 'XMLHttpRequest'
      });

    const output = JSON.parse(process.stdout.write.mock.calls[0][0]);
    const {headers} = output.meta.req;
    expect(headers['accept-language']).toBe('en-US,en;q=0.8,de;q=0.6,la;q=0.4');
    expect(headers['accept']).toBe('application/hal+json, application/json');
    expect(headers['x-requested-with']).toBe('XMLHttpRequest');
    expect(headers['cookie']).toBe('*****');
    expect(headers['authorization']).toBe('*****');

    logger.destroy();
  });

  afterEach(() => {
    process.stdout.write = originalStdout;
    process.stderr.write = originalStderr;
  });
});


src/createLoggerMiddleware.js
import {Router} from 'express';
import expressWinston from 'express-winston';
import bodyParser from 'body-parser';

expressWinston.responseWhitelist.push('body');

/**
 * Allows to integrate the logger with an express server.
 */

const MASKED_HEADERS = ['Cookie', 'cookie', 'Authorization', 'authorization'];
const MASKED_HEADER_VALUE = '*****';

export default ({logger}) => {
  const router = new Router();
  return router.use(
    bodyParser.json(),
    expressWinston.logger({
      winstonInstance: logger.winston,

      level(req, res) {
        if (res.statusCode >= 500) {
          return logger.getLevelsDescending()[0] || 'ERROR';
        }
        if (res.statusCode >= 400) {
          return logger.getLevelsDescending()[1] || 'WARN';
        }
        return logger.getLevelsDescending()[3] || 'DEBUG';
      },

      dynamicMeta(req) {
        const meta = {name: 'express'};

        if (req.method === 'POST' && req.body && req.body.operationName) {
          meta.graphql = {
            operationName: req.body.operationName,
            variables: req.body.variables
          };
        }

        return meta;
      },

      requestFilter(req, propName) {
        if (propName === 'headers') {
          // Mask confidential headers
          return Object.entries(req.headers)
            .map(([header, value]) => [
              header,
              MASKED_HEADERS.includes(header) ? MASKED_HEADER_VALUE : value
            ])
            .reduce((acc, [header, value]) => {
              acc[header] = value;
              return acc;
            }, {});
        }

        return req[propName];
      },

      responseFilter(res, propName) {
        if (propName === 'body') {
          const hasErrors =
            res.body && res.body.errors && res.body.errors.length > 0;

          // Ignore logging the body of responses, except for if a GraphQL error
          // happened.
          if (hasErrors) {
            return res[propName];
          }

          return;
        }

        return res[propName];
      }
    })
  );
};


src/Logger.js
import winston from 'winston';

/**
 * A small wrapper around winston for consistent configuration.
 */

export default class Logger {
  static defaultOpts = {
    colorize: process.env.NODE_ENV !== 'production',
    isProduction: process.env.NODE_ENV === 'production',

    // The higher the number, the higher the severity
    levels: {
      ERROR: 40000,
      WARN: 30000,
      INFO: 20000,
      DEBUG: 10000,
      TRACE: 5000
    },

    colors: {
      ERROR: 'red',
      WARN: 'yellow',
      INFO: 'green',
      DEBUG: 'blue',
      TRACE: 'cyan'
    },

    stderrLevels: ['ERROR']
  };

  constructor(opts = {}) {
    this.receiveOpts(opts);
    if (!this.service) throw new Error('`service` is mandatory');

    this.setupWinston();
    this.createLogMethods();
  }

  receiveOpts(opts) {
    const mergedOpts = {...Logger.defaultOpts, ...opts};
    Object.entries(mergedOpts).forEach(([key, val]) => {
      this[key] = val;
    });

    this.level = opts.level;
    if (!this.level) {
      if (this.isProduction) {
        this.level = this.getLevelsDescending().slice(-1)[0];
      } else {
        this.level = this.getLevelsDescending().slice(2)[0];
      }
    }
  }

  setupWinston() {
    // winston expects levels to conform to RFC 5424, which
    // specifies that lower levels correspond to higher severity.
    // To achieve the reverse order, a mapping is necessary.
    this.winstonLevels = this.getLevelsDescending().reduce((acc, cur, i) => {
      acc[cur] = i;
      return acc;
    }, {});

    this.winston = new winston.Logger({
      transports: [
        new winston.transports.Console({
          stderrLevels: this.stderrLevels,
          handleExceptions: false,
          colorize: this.colorize,
          formatter: this.isProduction
            ? (...args) => this.formatJSONMessage(...args)
            : undefined,
          level: this.level
        })
      ],
      levels: this.winstonLevels,
      colors: this.colors
    });

    // winston doesn't allow to set the error level for runtime
    // exceptions, therefore this should be handled outside of winston.
    // https://github.com/winstonjs/winston/blob/bd06c4091e9e248e6fed6278e0699e5cfdc7de24/lib/winston/exception-handler.js#L94-L97
    process.on('uncaughtException', this.onUnhandledException);

    // Unhandled promise rejections should be treated as errors.
    process.on('unhandledRejection', this.onUnhandledException);
  }

  createLogMethods() {
    Object.keys(this.levels).forEach(level => {
      this[level.toLowerCase()] = message => {
        let meta;

        // Log functions can be passed an object to provide
        // additional meta data like a logger `name`.
        if (typeof message === 'object') {
          meta = {...message};
          delete meta.message;
          message = message.message;
        }

        this.winston.log(level, message, meta);
      };
    });
  }

  onUnhandledException = e => {
    this.winston[this.getLevelsDescending()[0]](e.stack);
    process.exit(1);
  };

  getLevelsDescending() {
    return Object.entries(this.levels)
      .sort((a, b) => b[1] - a[1])
      .map(entry => entry[0]);
  }

  formatJSONMessage(opts) {
    const now = new Date();
    const {name: logger_name, ...meta} = opts.meta || {};
    const hasMeta = Object.keys(meta).length > 0;

    return JSON.stringify({
      service: this.service,
      '@timestamp': now.toISOString(),
      level: opts.level,
      level_value: this.levels[opts.level],
      logger_name,
      message: opts.message,
      meta: hasMeta ? meta : undefined,
      stack_trace: hasMeta && meta.stack ? meta.stack.join('\n') : undefined
    });
  }

  destroy() {
    process.removeListener('uncaughtException', this.onUnhandledException);
    process.removeListener('unhandledRejection', this.onUnhandledException);
    this.winston = null;
  }
}
